src/bin/mnemosyne/src/disk.rs

```diff
@@ -0,0 +1,170 @@
+use std::{
+    future,
+    mem::size_of,
+    sync::{Arc, Mutex, RwLock},
+    io::Error, io::ErrorKind
+};
+
+use std::io;
+
+use layout::{io::SeekFrom, ApplyLayout, Frame, Read, Seek, Write, IO};
+use crate::nvme::{init_nvme, NvmeController};
+
+use twizzler_async::block_on;
+
+use layout::{collections::raw::RawBytes, io::StdIO, Encode, SourcedDynamic};
+use lethe_gadget_fat::{
+    filesystem::FileSystem,
+    schema::{self, FATEntry, Superblock},
+};
+
+
+pub struct Disk {
+    ctrl: Arc<NvmeController>,
+    pub pos: usize,
+}
+
+impl Disk {
+    pub fn new() -> Result<Disk, ()> {
+        Ok(Disk {
+            ctrl: block_on(init_nvme()),
+            pos: 0
+        })
+    }
+}
+
+pub fn setup(data: &mut Disk) {
+    let super_block = Superblock {
+        magic: 0,
+        block_size: 0x200,
+        block_count: 0x1000,
+    };
+
+    let fat = vec![FATEntry::None; super_block.block_count as usize].into_boxed_slice();
+
+    let mut fs = schema::FileSystem {
+        super_block: super_block.clone(),
+        fat,
+        super_block_cp: super_block,
+        obj_lookup: vec![FATEntry::None; 3].into_boxed_slice(),
+        rest: RawBytes,
+    };
+
+    let fs_size = fs.sourced_size();
+    let reserved_blocks = fs_size / fs.super_block.block_size as u64
+        + (fs_size % fs.super_block.block_size as u64).min(1);
+
+    fs.fat[0] = FATEntry::Block(reserved_blocks);
+    fs.fat[1..reserved_blocks as usize].fill(FATEntry::Reserved);
+    for i in reserved_blocks..fs.super_block.block_count - 1 {
+        fs.fat[i as usize] = FATEntry::Block(i + 1);
+    }
+    fs.fat[fs.super_block.block_count as usize - 1] = FATEntry::None;
+
+    fs.encode(data).unwrap();
+}
+
+const DISK_SIZE: usize = 0x1000000;
+const PAGE_SIZE: usize = 4096;
+const SECTOR_SIZE: usize = 512;
+const PAGE_SHIFT: usize = 12;
+const PAGE_MASK: usize = 0xFFF;
+const LBA_COUNT: usize = DISK_SIZE / SECTOR_SIZE;
+
+impl Read for Disk {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, Self::Error> {
+        let mut lba = (self.pos / PAGE_SIZE) * 8;
+        let mut bytes_written: usize = 0;
+        let mut read_buffer: [u8; PAGE_SIZE] = [0; PAGE_SIZE];
+
+        while bytes_written != buf.len() {
+            if lba >= LBA_COUNT {
+                break;
+            }
+
+            let left = self.pos % PAGE_SIZE;
+            let right = if buf.len() - bytes_written > PAGE_SIZE {PAGE_SIZE} else {left + buf.len() - bytes_written}; // If I want to write more than the boundary of a page
+            block_on(self.ctrl.read_page(lba as u64, &mut read_buffer, 0));
+
+            for p in left..right {
```

Comment:
I'd suggest using [copy_from_slice](https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice) and then adding the len to bytes_written.